packages/realm/src/schema/validate.ts

```diff
@@ -73,50 +74,58 @@ export function validateRealmSchema(realmSchema: unknown): asserts realmSchema i
+        }
       }
+      assert.object(clazz.schema, "schema static");
+      validateObjectSchema(clazz.schema);
     }
-    assert.object(clazz.schema, "schema static");
-    validateObjectSchema(clazz.schema);
-  }
-  // Schema is passed as an object (ObjectSchema)
-  else {
-    assert.object(objectSchema, "object schema", { allowArrays: false });
-    const { name: objectName, properties, primaryKey, asymmetric, embedded } = objectSchema;
-    assert.string(objectName, "'name' on object schema");
-    assert.object(properties, `'properties' on '${objectName}'`, { allowArrays: false });
-    if (primaryKey !== undefined) {
-      assert.string(primaryKey, `'primaryKey' on '${objectName}'`);
-    }
-    if (embedded !== undefined) {
-      assert.boolean(embedded, `'embedded' on '${objectName}'`);
-    }
-    if (asymmetric !== undefined) {
-      assert.boolean(asymmetric, `'asymmetric' on '${objectName}'`);
-    }
+    // Schema is passed as an object (ObjectSchema)
+    else {
+      assert.object(objectSchema, "object schema", { allowArrays: false });
+      const { name: objectName, properties, primaryKey, asymmetric, embedded } = objectSchema;
+      assert.string(objectName, "'name' on object schema");
+      assert.object(properties, `'properties' on '${objectName}'`, { allowArrays: false });
+      if (primaryKey !== undefined) {
+        assert.string(primaryKey, `'primaryKey' on '${objectName}'`);
+      }
+      if (embedded !== undefined) {
+        assert.boolean(embedded, `'embedded' on '${objectName}'`);
+      }
+      if (asymmetric !== undefined) {
+        assert.boolean(asymmetric, `'asymmetric' on '${objectName}'`);
+      }
 
-    const invalidKeysUsed = filterInvalidKeys(objectSchema, OBJECT_SCHEMA_KEYS);
-    assert(
-      !invalidKeysUsed.length,
-      `Unexpected field(s) found on the schema for object '${objectName}': '${invalidKeysUsed.join("', '")}'.`,
-    );
+      const invalidKeysUsed = filterInvalidKeys(objectSchema, OBJECT_SCHEMA_KEYS);
+      assert(
+        !invalidKeysUsed.length,
+        `Unexpected field(s) found on the schema for object '${objectName}': '${invalidKeysUsed.join("', '")}'.`,
+      );
 
-    for (const propertyName in properties) {
-      const propertySchema = properties[propertyName];
-      const isUsingShorthand = typeof propertySchema === "string";
-      if (!isUsingShorthand) {
-        validatePropertySchema(objectName, propertyName, propertySchema);
+      for (const propertyName in properties) {
+        const propertySchema = properties[propertyName];
+        const isUsingShorthand = typeof propertySchema === "string";
+        if (!isUsingShorthand) {
+          validatePropertySchema(objectName, propertyName, propertySchema);
+        }
       }
     }
+  } catch (err: any) {
+    // Rethrow as SchemaParseError rather than a mix of Error, TypeError,
+    // TypeAssertionError, or AssertionError.
+    throw new SchemaParseError(err.message);
   }
```

Comment:
I believe this pattern is better as it won't risk getting the `message` off something that wasn't an object.

```suggestion
  } catch (err) {
    if (err instanceof Error) {
      // Rethrow as SchemaParseError rather than a mix of Error, TypeError,
      // TypeAssertionError, or AssertionError.
      throw new SchemaParseError(err.message);
    } else {
      throw err;
    }
  }
```